### ISSUE TITLE
I am unable to convert a DateTime.Now variable ToString("H")

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#49471]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity2.visualstudio.com/t/I-am-unable-to-convert-a-DateTimeNow-va/1258838?preview=true)._

---
[severity:It's more difficult to complete my work]
I am defining a variable as such:

DateTime today = DateTime.Now;

string hour = today. ToString("H")

This causes an exception/error. I can put another character in and it will output the 24-clock hour just fine.

---
#### Feedback Bot on 11/17/2020, 09:40 PM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

---
### Original Solutions
(no solutions)

**Comment by sharwell** [comment]
@stephentoub this behavior occurs because a user is being (reasonably) misled by our DateTime format string IntelliSense feature:

![image](https://user-images.githubusercontent.com/1408396/99569122-18e54580-2985-11eb-88ac-e002452df395.png)


I'm curious if you have suggestions for how we handle this.

/cc @CyrusNajmabadi

**Comment by stephentoub** [comment]
Related: https://github.com/dotnet/runtime/issues/12192

**Comment by CyrusNajmabadi** [comment]
> I'm curious if you have suggestions for how we handle this.

We likely should update the doc string here to explain this.

**Comment by sharwell** [comment]
@CyrusNajmabadi I believe the embedded languages feature also supports diagnostics, so we could show a warning for these unsupported formats

**Comment by CyrusNajmabadi** [comment]
Ah, we do show this in the format string:

> If the "h" format specifier is used without other custom format specifiers, it's interpreted as a standard date and time format specifier and throws a FormatException.

> I believe the embedded languages feature also supports diagnostics, so we could show a warning for these unsupported formats

I would be ok with this.  Though it could also be provided by runtime.

**Comment by sharwell** [comment]
> Though it could also be provided by runtime.

Agreed here. @stephentoub is there a tracking issue for an SDK analyzer covering calls to BCL implementations of `ToString(string)`?

**Comment by stephentoub** [comment]
> Agreed here. @stephentoub Stephen Toub FTE is there a tracking issue for an SDK analyzer covering calls to BCL implementations of ToString(string)?

Not to my knowledge, but an analyzer helping to validate format specifiers are being used correctly sounds very reasonable.
cc: @terrajobst

**Comment by MaStr11** [comment]
>> I believe the embedded languages feature also supports diagnostics, so we could show a warning for these unsupported formats

There is `AbstractRegexDiagnosticAnalyzer` which reports "RE0001" and could be used as a template. https://github.com/dotnet/roslyn/blob/1f97e38e25c1bfd80728d544a4252f94ee95efbc/src/Features/Core/Portable/EmbeddedLanguages/RegularExpressions/AbstractRegexDiagnosticAnalyzer.cs#L20-L26

> I would be ok with this. Though it could also be provided by runtime.

Is there any decision yet, where the analyzer should go?

**Comment by MaStr11** [comment]
> @stephentoub is there a tracking issue for an SDK analyzer covering calls to BCL implementations of ToString(string)?

I created one dotnet/runtime#45960

**Comment by CyrusNajmabadi** [comment]
Closing out as this is tracked over at https://github.com/dotnet/runtime/issues/45960
